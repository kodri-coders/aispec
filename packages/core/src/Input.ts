import { type XMLBaseConstructor } from './XMLBase';

interface IInputConstructor extends XMLBaseConstructor {
  name: string;
  description: string;
  schema: string;
}

export class Input {
  name: string;
  description: string;
  schema: string;

  constructor(public input: IInputConstructor) {
    this.name = input.name;
    this.description = input.description;
    this.schema = JSON.parse(input.schema);
  }
};
