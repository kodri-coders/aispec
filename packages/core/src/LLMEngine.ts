import { generateText, tool } from 'ai';
interface Model {
  name: string;
  temperature: number;
  max_tokens: number;
}
export class LLMEngine {
  model: Model;
  constructor(model: Model){
    this.model = model;
  }
  modelFactory(model: Model): LLMEngine {
    return new LLMEngine(model);
  }
  prepareRequest({
    systemPrompt,
    prompt,
    tools,
    model,
  }: {
    systemPrompt: string;
    prompt: string;
    tools: any;
    model: any;
  }): any {
    switch (model.name) {
      case 'gpt-4':
      case 'gpt-4o':
      case 'gpt-3.5':
        return {
          system: systemPrompt,
          prompt,
          tools,
          model: model.name,
          temperature: model.temperature,
          maxTokens: model.max_tokens,
        };
      case 'o1-mini':
      case 'o1-preview':
        return {
          messages: [
            {
              role: 'user',
              content: systemPrompt
            },
            {
              role: 'user',
              content: prompt
            }
          ],
          tools,
          model: model.name,
        };
      default:
        throw new Error('Model not supported');
    }
  }
  async generateText({
    prompt,
    systemPrompt,
  }: {
    prompt: string;
    systemPrompt: string;
  }): Promise<string> {
    const response = await generateText(
      this.prepareRequest({
        systemPrompt,
        prompt,
        tools: [],
        model: this.model,
      }));
    return response.text;
  }
}