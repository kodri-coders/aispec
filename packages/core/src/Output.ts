import { type XMLBaseConstructor } from './XMLBase';

interface IOutputConstructor extends XMLBaseConstructor {
  description: string;
  name: string;
  schema: string;
}

export class Output {
  description: string;
  name: string;
  schema: string;
  node: IOutputConstructor;

  constructor(public output: IOutputConstructor) {
    this.name = output.name;
    this.description = output.description;
    this.node = output;
    try {
      if (typeof output.schema === 'object') {
        this.schema = JSON.parse(output.schema['#text']);
      }
      else {
        this.schema = JSON.parse(output.schema);
      }
    }
    catch (error) {
      console.error(error);
      throw new Error('Invalid output schema');
    }
  }
};
