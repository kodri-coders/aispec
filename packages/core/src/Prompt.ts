export class Prompt {
  public variables: string[];
  private prompt: string;

  constructor(prompt: string) {
    this.prompt = prompt;
    this.variables = this.extractVariables(prompt);
  }

  extractVariables(prompt?: string): string[] {
    const variables = (prompt || this.prompt).match(/\$\{(.+?)\}/g);
    return (variables || []).map(variable => variable.replace(/\${|}/g, ''));
  }

  interpolate(data: any): string {
    return this.prompt.replace(/\${(.+?)}/g, (_, key) => {
      let value = '()';
      key.split('.').forEach((keyPart: string) => {
        value = data[keyPart];
      });
      return value;
    });
  }
};
