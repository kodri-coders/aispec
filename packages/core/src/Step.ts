import { Input } from './Input';
import { Prompt } from './Prompt';
import { XMLBase, type XMLBaseConstructor } from './XMLBase';
import { Assistant } from './Assistant';
import { Output } from './Output';

interface IStepConstructor extends XMLBaseConstructor {
  name: string;
  description: string;
  input: any;
  prompt: string;
  outputs: any;
  $ref?: string;
}

interface StepInterface {
  name: string;
  description: string;
  output: Output;
  prompt: Prompt;
  excludedTags: string[];
} 
export class Step extends XMLBase {
  name?: string;
  description?: string = '';
  output?: Output;
  prompt?: Prompt;
  excludedTags: string[] = ['model'];
  mapping: any= {
    id: "@id",
    name: 'name',
    description: 'description',
    input: ['input', Input],
    prompt: ['prompt', Prompt],
    output: ['output', Output],
    model: 'model'
  }
  constructor(
    public step: IStepConstructor,
  ) {
    super(step);
    this.mount();
  }

  getPrompt(input?: any): any {
    return this.prompt?.interpolate(input);
  }
};
