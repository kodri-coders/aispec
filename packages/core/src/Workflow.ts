import { Assistant } from './Assistant';
import { Step } from './Step';
import { XMLBase, type XMLBaseConstructor } from './XMLBase';

interface IWorkflowConstructor extends XMLBaseConstructor {
  '@id': string;
  name: string;
  description: string;
  steps: {
    step: {
      name: string;
      description: string;
      input: any;
      prompt: string;
      outputs: any;
    }[];
  };
}

export class Workflow extends XMLBase {
  name: string='';
  description: string='';
  steps: Step[] = [];
  excludedTags: string[] = ['model'];
  mapping: any= {
    id: "@id",
    name: 'name',
    description: 'description',
    steps: ['steps.step', Step],
    model: 'model'
  }
  constructor(
    public workflow: IWorkflowConstructor,
  ) {
    super(workflow);
    this.mount();
  }

  run(input?: any): any {
    return this.steps.map((step: any) => step.runStep(input));
  }
};
