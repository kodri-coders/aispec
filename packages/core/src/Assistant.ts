import { Skill } from './Skill';
import { Workflow } from './Workflow';
import { WorkflowRunner } from './WorkflowRunner';
import { XMLBase, type XMLBaseConstructor } from './XMLBase';

interface IAssistantConstructor extends XMLBaseConstructor {
  '@id'?: string;
  name?: string;
  description?: string;
  skills?: {
    skill: any[];
  };
  workflows?: {
    workflow: any[];
  };
}

export class Assistant extends XMLBase {
  name?: string;
  description?: string;
  workflows?: Workflow[] = [];
  skills?: Skill[] = [];

  constructor(assistant: IAssistantConstructor | XMLBase) {
    super(assistant);
    this.name = this.node.name;
    this.description = this.node.description;
    this.model = this.node.model;

    this.skills = this.node.skills?.skill.map((skill: any) => {
      return new Skill(skill, this);
    });

    this.workflows = this.node.workflows?.workflow.map((workflow: any) => {
      return new Workflow(workflow);
    });
  }

  findWorkflow(workflowId: string): any {
    let targetWorkflow: Workflow | undefined;
    let containingSkill: Skill | undefined;

    // Check global workflows
    targetWorkflow = this.workflows?.find(
      (w) => w.workflow['@id'] === workflowId,
    );

    // If not found in global workflows, check skills
    if (!targetWorkflow) {
      this.skills?.forEach((skill) => {
        const found = skill.workflows?.find(
          (w) => w.workflow['@id'] === workflowId,
        );
        if (found) {
          targetWorkflow = found;
          containingSkill = skill;
        }
      });
    }

    if (!targetWorkflow) {
      throw new Error(`Workflow with id ${workflowId} not found`);
    }
    return { workflow: targetWorkflow, skill: containingSkill };
  }

  loadWorkflow(workflowId: string, input?: any): any {
    // Find the workflow and its containing skill
    const { workflow: targetWorkflow, skill: containingSkill } =
      this.findWorkflow(workflowId);
    const xmlObj: any = {
      assistant: {
        name: this.name,
        description: this.description,
        skills: {
          skill: [],
        },
      },
    };

    // Add global skills
    if (this.skills) {
      xmlObj.assistant.skills.skill = this.skills
        .filter((s) => s !== containingSkill)
        .map((s) => ({ name: s.name, description: s.description }));
    }
    if (targetWorkflow) {
      xmlObj.assistant.workflows = {
        workflow: [targetWorkflow],
      };
    }
    // Add the skill containing the workflow
    if (containingSkill) {
      xmlObj.assistant.skills.skill.push(containingSkill.node);
    }

    return new WorkflowRunner(new Assistant(xmlObj), targetWorkflow);
  }
}
