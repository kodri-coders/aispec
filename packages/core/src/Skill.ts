import { Assistant } from './Assistant';
import { Workflow } from './Workflow';
import { XMLBase, type XMLBaseConstructor } from './XMLBase';

interface ISkillConstructor extends XMLBaseConstructor {
  '@id': string;
  name: string;
  description: string;
  workflows?: {
    workflow: any[];
  };
}

export class Skill extends XMLBase {
  workflows?: Workflow[] = [];
  excludedTags: string[] = ['model'];
  skills?: Skill[] = [];
  mapping: any= {
    id: "@id",
    workflows: ['workflows.workflow', Workflow],
    skills: ['skills.skill', Skill],
    description: 'description',
    name: 'name'
  }
  constructor(
    public skill: ISkillConstructor,
  ) {
    super(skill);
    this.mount();
  }
};
