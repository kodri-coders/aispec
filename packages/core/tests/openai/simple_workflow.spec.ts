import { Assistant } from "../../src";
import * as path from "path";

const artifactPath = path.join(
  __dirname,
  "artifacts",
  "plant_finder.xml",
);
const assistant = new Assistant({ "@ref": artifactPath });
describe("assistant", () => {
  it("loads name and description", () => {
    expect(assistant.name).toMatchSnapshot();
    expect(assistant.description).toMatchSnapshot();
  });
  it("loads workflows", () => {
    expect(assistant.workflows?.length).toMatchSnapshot();
  });
  it("starts workflow", async () => {
    const workflowRunner = assistant.loadWorkflow("plant-find");
    workflowRunner.on("step-finished", (prompt: string, data: any) => {
      console.log("step finished");
      workflowRunner.next();
    });
    workflowRunner.start({
      plant_query: "Tropical plants that grow in Athens, Greece on a pot in my balcony",
    });

    await workflowRunner.onFinish();
    expect(workflowRunner.context).toMatchSnapshot();
    expect(workflowRunner.logs).toMatchSnapshot();
  },100000);
});
