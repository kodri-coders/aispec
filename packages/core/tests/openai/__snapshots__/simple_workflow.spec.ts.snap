// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assistant loads name and description 1`] = `"Plant finder"`;

exports[`assistant loads name and description 2`] = `"Find plants"`;

exports[`assistant loads workflows 1`] = `1`;

exports[`assistant starts workflow 1`] = `
{
  "plant_details": [
    {
      "plant": {
        "description": "The Bird of Paradise is a striking tropical plant known for its vibrant, bird-like flowers. Native to South Africa, it is a popular ornamental plant used in gardens and floral arrangements. The plant has large, banana-like leaves and can reach considerable heights. It thrives in warm climates and requires well-drained soil and full sun exposure to flourish.",
        "latin_name": "Strelitzia reginae",
        "name": "Bird of Paradise",
      },
    },
    {
      "plant": {
        "description": "Ficus elastica, commonly known as the Rubber Plant, is a species of plant in the Moraceae family. Native to eastern parts of South Asia and Southeast Asia, it can grow as a large evergreen tree in its native habitat, reaching up to 30 meters (100 feet) tall. As a houseplant, it typically grows between 2.5 to 6 meters (8 to 20 feet). The Rubber Plant is known for its broad, glossy, dark green leaves, which can develop a burgundy tint under the right conditions. It's popular in interior design for its aesthetic appeal and relative ease of care, requiring moderate light, watering, and humidity levels.",
        "latin_name": "Ficus elastica",
        "name": "Rubber Plant",
      },
    },
    {
      "plant": {
        "description": "Philodendrons are a large genus of flowering plants in the Araceae family, native to the tropical Americas. They are known for their large, often heart-shaped leaves and vining or climbing growth habit. Philodendrons are popular as houseplants due to their attractive foliage and adaptability to indoor conditions.",
        "latin_name": "Philodendron spp.",
        "name": "Philodendron",
      },
    },
    {
      "plant": {
        "description": "Monstera deliciosa, also known as the Swiss cheese plant or split-leaf philodendron, is a popular tropical plant native to the rainforests of southern Mexico to Panama. It is characterized by its large, glossy green leaves that develop natural splits and holes as they mature, giving it a distinctive appearance. It thrives in bright, indirect light and requires well-draining soil with regular watering, allowing the top inch of soil to dry out between waterings. Monstera deliciosa is valued for its ornamental foliage and is commonly grown as a houseplant in temperate climates.",
        "latin_name": "Monstera deliciosa",
        "name": "Monstera deliciosa",
      },
    },
    {
      "plant": {
        "description": "Dracaena is a genus of about 120 species of trees and succulent shrubs, commonly known as dracaenas. They are popular as houseplants due to their attractive foliage and low maintenance requirements. Dracaenas come in various sizes and leaf shapes, and some species are known for their air-purifying qualities.",
        "latin_name": "Dracaena",
        "name": "Dracaena",
      },
    },
    {
      "plant": {
        "description": "Codiaeum variegatum, commonly known as croton, is a vibrant houseplant with colorful, variegated leaves. It is native to Southeast Asia and is popular for its bright foliage that comes in various colors including green, yellow, red, and orange. Crotons require bright, indirect light and regular watering, making them a favorite for indoor gardening enthusiasts.",
        "latin_name": "Codiaeum variegatum",
        "name": "Croton",
      },
    },
  ],
  "plant_query": "Tropical plants that grow in Athens, Greece on a pot in my balcony",
  "plants": [
    "Bird of Paradise",
    "Ficus elastica",
    "Philodendron",
    "Monstera deliciosa",
    "Dracaena",
    "Croton",
  ],
}
`;
